linera-views/src/indexed_db.rs

```diff
@@ -0,0 +1,359 @@
+        let namespace = namespace.to_string();
+        let object_store_name = namespace.clone();
+        let mut database = IdbDatabase::open(DATABASE_NAME)?.await?;
+
+        if !database.object_store_names().any(|n| n == namespace) {
+            let version = database.version();
+            database.close();
+            let mut db_req = IdbDatabase::open_f64(DATABASE_NAME, version + 1.0)?;
+            db_req.set_on_upgrade_needed(Some(move |event: &IdbVersionChangeEvent| {
+                event.db().create_object_store(&namespace)?;
+                Ok(())
+            }));
+            database = db_req.await?;
+        }
+
+        Ok(IndexedDbStore {
+            database,
+            object_store_name,
+            max_stream_queries: config.common_config.max_stream_queries,
+        })
+    }
+
+    async fn list_all(config: &Self::Config) -> Result<Vec<String>, IndexedDbContextError> {
+        Ok(Self::connect(config, "")
+            .await?
+            .database
+            .object_store_names()
+            .collect())
+    }
+
+    async fn exists(config: &Self::Config, namespace: &str) -> Result<bool, IndexedDbContextError> {
+        Ok(Self::connect(config, "")
+            .await?
+            .database
+            .object_store_names()
+            .any(|x| x == namespace))
+    }
+
+    async fn create(config: &Self::Config, namespace: &str) -> Result<(), IndexedDbContextError> {
+        Self::connect(config, "")
+            .await?
+            .database
+            .create_object_store(namespace)?;
+        Ok(())
+    }
+
+    async fn delete(config: &Self::Config, namespace: &str) -> Result<(), IndexedDbContextError> {
+        Ok(Self::connect(config, "")
+            .await?
+            .database
+            .delete_object_store(namespace)?)
+    }
+}
+
+impl LocalKeyValueStore for IndexedDbStore {
+    type Error = IndexedDbContextError;
+}
+
+/// An implementation of [`crate::common::Context`] that stores all values in memory.
+pub type IndexedDbContext<E> = ContextFromStore<E, IndexedDbStore>;
+
+impl<E> IndexedDbContext<E> {
+    /// Creates a [`IndexedDbContext`].
+    pub fn new(store: IndexedDbStore, extra: E) -> Self {
+        Self {
+            store,
+            base_key: vec![],
+            extra,
+        }
+    }
+}
+
+/// Provides a `IndexedDbContext<()>` that can be used for tests.
+pub async fn create_indexed_db_test_context() -> IndexedDbContext<()> {
+    IndexedDbContext::new(
+        create_indexed_db_store_stream_queries(TEST_INDEX_DB_MAX_STREAM_QUERIES).await,
+        (),
+    )
+}
+
+/// Creates a test memory client for working.
+pub async fn create_indexed_db_store_stream_queries(max_stream_queries: usize) -> IndexedDbStore {
+    let common_config = CommonStoreConfig {
+        max_concurrent_queries: None,
+        max_stream_queries,
+        cache_size: 1000,
+    };
+    let config = IndexedDbStoreConfig { common_config };
```

Comment:
That code seems to duplicate with the code above.

Reply:
Ah yes — I took it from `memory.rs`, but you're right, it is duplicated :)  Thanks for the spot!